provisioning/e2e/_provisioning_e2e.js

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+'use strict';
+
+var Promise = require('bluebird');
+var pem = Promise.promisifyAll(require('pem'));
+var uuid = require('uuid');
+var assert = require('chai').assert;
+var debug = require('debug')('azure-device-provisioning-e2e');
+var Http = require('azure-iot-provisioning-device-http').Http;
+var ProvisioningServiceClient = require('azure-iot-provisioning-service').ProvisioningServiceClient;
+var Registry = require('azure-iothub').Registry;
+
+var idScope = process.env.IOTHUB_PROVISIONING_IDSCOPE;
+var provisioningConnectionString = process.env.IOTHUB_PROVISIONING_CONNECTION_STRING;
+var registryConnectionString = process.env.IOTHUB_CONNECTION_STRING;
+
+var provisioningTransports = [ Http ];
+
+describe('IoT Provisioning', function() {
+  var deviceId;
+  var registrationId;
+  this.timeout(30000);
+
+  var provisioningServiceClient = Promise.promisifyAll(ProvisioningServiceClient.fromConnectionString(provisioningConnectionString));
+  var registry = Promise.promisifyAll(Registry.fromConnectionString(registryConnectionString));
+
+  var createX509Cert = function(registrationId) {
+    var certOptions = {
+      commonName: registrationId,
+      selfSigned: true,
+      days: 10
+    };
+
+    return pem.createCertificateAsync(certOptions)
+      .then(function(result) {
+        return {
+          cert: result.certificate,
+          key: result.clientKey
+        };
+      });
+  };
+
+  var enrollX509 = function(registrationId, deviceId, x509) {
+    var enrollment =  {
+      registrationId: registrationId,
+      deviceId: deviceId,
+      attestation: {
+        type: 'x509',
+        x509: {
+          clientCertificates: {
+            primary: {
+              certificate: x509.cert
+            }
+          }
+        }
+      },
+      initialTwinState: {
+        desiredProperties: {
+          testProp: registrationId + ' ' + deviceId
+        }
+      }
+    };
+
+    return provisioningServiceClient.createOrUpdateIndividualEnrollmentAsync(enrollment);
+  };
+
+  beforeEach (function() {
+    var id = uuid.v4();
+    deviceId = 'deleteme_provisioning_node_e2e_' + id;
+    registrationId = 'reg-' + id;
+  });
+
+  afterEach (function(callback) {
+    debug('deleting enrollment');
+    provisioningServiceClient.deleteIndividualEnrollmentAsync(registrationId)
+      .catch(function() {
```

Comment:
>catch [](start = 7, length = 5)

you should be able to use only one catch statement at the end after all the then() operations - this being said, not sure if promises serve our purpose here.

Reply:
Like I said, I go back and forth on the usefulness of promises.

However, this catch exists to ignore the error on deleteIndividualEnrollment then continue.  In other words, if deleteIndividualEnrollment fails, this catch will execute, then the flow will continue with a call to registry.delete.

reference: [http://odetocode.com/blogs/scott/archive/2015/10/01/javascript-promises-and-error-handling.aspx](http://odetocode.com/blogs/scott/archive/2015/10/01/javascript-promises-and-error-handling.aspx)

'Just like normal execution can resume in procedural code after a catch statement, normal execution can resume with promises after a handled error.'

---
In reply to: [148917887](https://github.com/Azure/azure-iot-sdk-node/pull/135#discussion_r148917887) [](ancestors = 148917887)